crates/holochain/src/core/workflow/countersigning_workflow.rs

```diff
@@ -56,180 +56,83 @@ impl CountersigningWorkspace {
-    /// a subset of the network, but it does protect us against Holochain forcing a decision while
-    /// it is unable to reach any peers.
-    ///
-    /// Note that because the [PreflightRequest] is stored here, we only ever enter the unknown state
-    /// if we managed to keep the preflight request in memory, or if we have been able to recover it
-    /// from the source chain as part of the committed [CounterSigningSessionData]. Otherwise, we
-    /// are unable to discover what session we were participating in, and we must abandon the session
-    /// without going through this recovery state.
-    Unknown {
-        preflight_request: PreflightRequest,
-        resolution: SessionResolutionSummary,
-    },
-}
 
-impl CountersigningSessionState {
-    fn preflight_request(&self) -> &PreflightRequest {
-        match self {
-            CountersigningSessionState::Accepted(preflight_request) => preflight_request,
-            CountersigningSessionState::SignaturesCollected {
-                preflight_request, ..
-            } => preflight_request,
-            CountersigningSessionState::Unknown {
-                preflight_request, ..
-            } => preflight_request,
-        }
+    pub fn get_countersigning_session_state(&self) -> Option<CountersigningSessionState> {
+        self.inner
+            .share_ref(|inner| Ok(inner.session.clone()))
+            .unwrap()
     }
 
-    fn session_app_entry_hash(&self) -> &EntryHash {
-        let request = match self {
-            CountersigningSessionState::Accepted(request) => request,
-            CountersigningSessionState::SignaturesCollected {
-                preflight_request, ..
-            } => preflight_request,
-            CountersigningSessionState::Unknown {
-                preflight_request, ..
-            } => preflight_request,
-        };
-
-        &request.app_entry_hash
+    pub fn mark_countersigning_session_for_force_abandon(
+        &self,
+        cell_id: &CellId,
+    ) -> Result<(), CountersigningError> {
+        self.inner
+            .share_mut(|inner, _| {
+                Ok(if let Some(session) = inner.session.as_mut() {
+                    if let CountersigningSessionState::Unknown {
+                        resolution,
+                        force_abandon,
+                        ..
+                    } = session
+                    {
+                        if resolution.attempts >= 1 {
+                            *force_abandon = true;
+                            Ok(())
+                        } else {
+                            Err(CountersigningError::SessionNotUnresolved(cell_id.clone()))
+                        }
+                    } else {
+                        Err(CountersigningError::SessionNotUnresolved(cell_id.clone()))
+                    }
+                } else {
+                    Err(CountersigningError::SessionNotFound(cell_id.clone()))
+                })
+            })
+            .unwrap()
     }
```

Comment:
These two calls are quite wordy. I couldn't think of a more concise form or how to extract something shared.